bridge/src/database.rs

```diff
@@ -39,30 +46,156 @@ impl fmt::Display for Database {
+	/// `Database` containing its parsed contents.
+	pub fn load_user_defined<P: AsRef<Path>>(path: P, home_contract_address: Address,
+			foreign_contract_address: Address)
+			-> Result<Database, Error> {
 		let mut file = match fs::File::open(&path) {
 			Ok(file) => file,
-			Err(ref err) if err.kind() == io::ErrorKind::NotFound => return Err(ErrorKind::MissingFile(format!("{:?}", path.as_ref())).into()),
-			Err(err) => return Err(err).chain_err(|| "Cannot open database"),
+			Err(ref err) if err.kind() == io::ErrorKind::NotFound =>
+				return Err(ErrorKind::MissingFile(format!("{:?}", path.as_ref())).into()),
+			Err(err) => return Err(err).chain_err(|| "Cannot open database file"),
 		};
 
 		let mut buffer = String::new();
 		file.read_to_string(&mut buffer)?;
-		buffer.parse()
+		Database::from_str_user_defined(&buffer, home_contract_address, foreign_contract_address)
 	}
 
-	pub fn save<W: Write>(&self, mut write: W) -> Result<(), Error> {
-		write.write_all(self.to_string().as_bytes())?;
+	/// Returns a new `Database` constructed from a stored toml string
+	/// containing keys for 'home_contract_address' and
+	/// 'foreign_contract_address'.
+	#[cfg(test)]
+	fn from_str_stored<S: AsRef<str>>(s: S) -> Result<Database, Error> {
+		toml::from_str::<parsed::StoredDatabase>(s.as_ref())
+			.map(Database::from)
+			.map_err(Error::from)
+	}
+
+	/// Returns a new `Database` constructed from the parsed string `s` and
+	/// the provided addresses.
+	///
+	/// An error will be returned if the `s` contains keys for
+	/// 'home_contract_address' or 'foreign_contract_address'.
+	fn from_str_user_defined<S: AsRef<str>>(s: S, home_contract_address: Address,
+			foreign_contract_address: Address) -> Result<Database, Error> {
+		let db_parsed: parsed::UserDefinedDatabase = toml::from_str(s.as_ref())
+			.chain_err(|| "Cannot parse database file")?;
+
+		Ok(Database {
+			home_contract_address,
+			foreign_contract_address,
+			home_deploy: db_parsed.home_deploy,
+			foreign_deploy: db_parsed.foreign_deploy,
+			checked_deposit_relay: db_parsed.checked_deposit_relay,
+			checked_withdraw_relay: db_parsed.checked_withdraw_relay,
+			checked_withdraw_confirm: db_parsed.checked_withdraw_confirm,
+		})
+	}
+
+	/// Writes a serialized `Database` to a writer.
+	#[deprecated(note = "Use '::store' instead.")]
+	pub fn save<W: Write>(&self, writer: W) -> Result<(), Error> {
+		self.store(writer)
+	}
+
+	/// Writes a serialized `Database` to a writer.
+	pub fn store<W: Write>(&self, mut writer: W) -> Result<(), Error> {
+		writer.write_all(self.to_string().as_bytes())?;
 		Ok(())
 	}
 }
 
+mod parsed {
+	#[cfg(test)]
+	use std::fmt;
+	#[cfg(test)]
+	use toml;
+
+	use super::Address;
+
+	/// Parsed application "database".
+	#[derive(Debug, Deserialize, Serialize)]
+	#[serde(deny_unknown_fields)]
+	pub struct StoredDatabase {
+		/// Address of home contract.
+		pub home_contract_address: Address,
```

Comment:
I think that I'm confused about when we will use the config object to retrieve the contract addresses and when we will use the database. Is having the home and foreign contract addresses in both objects redundant?

Reply:
The database configurations, from what I understand, may possibly be stored in a previously serialized format (via the `Database::save/store` method). I'm not sure exactly how they're used in practice but I imagine it's convenient to be able to back up a database configuration as a whole unit.

If its not necessary to store database configs containing contract addresses, and there is a guarantee to always have a good configuration loaded from the `config.toml`, I can remove that functionality entirely. In which case we could probably also remove the `Option` from the `Note::contract_addresses` as well.

Reply:
Everything in the PR is really good so far.

> from what I understand, may possibly be stored in a previously serialized format 

This is a good point, but I would assume anyone who currently is using a `db.toml` file and is regularly updating their `poa-bridge` instance with `master`, will be able to change he contract address fields from their database file to their config file.

From the line of thinking in issue #101 that:

- A database is used for persistent storage of block numbers that have been
checked for various contract events. The bridge has access to write to this
file during runtime.

- The config file is read only, and should contain everything else.

We could probably do the following:

- We should remove `home_contract_address` and `foreign_contract_address` from the database structures in `database.rs`.

- `parsed::Node` (in `config.rs`) should have the field: `contract_address: Option<Address>`,

- `Node` (in `config.rs`) should have a field `contract_address: Address`.

- `Node::from_load_struct()` should contain something like the following:

```rust
let contract_address = match node.contract_address {
    Some(addr) => addr,
    None => {
        let msg =
            "Config file is missing contract address. Please define a \
            `contract_address` for both `[home]` and `[foreign]`.".into();
        
        return Err(ErrorKind::ConfigError(msg).into());
    },
};
...

let node = Node {
    ...
    contract_address,
    ...
};
```

and this block could be removed:

```rust
if cfg!(not(feature = "deploy")) && node.contract_address.is_none() {
    ...
}
```

- In `desposit_relay.rs`, `withdraw_relay.rs`, and `withdraw_confirm.rs` change any `init.home/foreign_contract_address` -> `app.config.home/foreign.contract_address`.

Reply:
Ok great, that clarifies things.

Forgive my ignorance but could you explain to me when `deploy` builds are used and what the usage under that configuration entails? I don't quite have a clear picture of the entire ecosystem and I haven't seen any documentation about it.

Also, do we need to worry about handling stored configurations from the prior (current) version of `poa-bridge` in an especially user-friendly way? Usually when I make a configuration file change to a program I leave all sorts of friendly error messages, etc. to ease the transition. How much of that is necessary here? Are the 'stored', serialized configuration files loaded in by the `Deploy` struct, for example, kept somewhere persistently or are they generated from scratch each time the bridge is started? Please point me to any material that might help me understand how the bridge fits in to its surrounding software elements.

I'll try out the changes you've suggested this afternoon or evening. Thanks a lot for your help :)

Reply:
@c0gent Nick, as for the `deploy` feature, it is on its way out as it is a remnant of the original version that used to do the deployment itself. It is currently only used by integration tests and this will removed with the transition to v2 of contracts (#103)

Reply:
I see. That simplifies things a bit. Thank you.

Reply:
I'll push my latest changes incorporating these ideas sometime tomorrow afternoon.